org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/DecompilerManager.java

```diff
@@ -0,0 +1,192 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     David Gileadi - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.ls.core.internal.managers;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.ISafeRunnable;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.core.runtime.SafeRunner;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.jdt.core.IClassFile;
+import org.eclipse.jdt.ls.core.internal.Disassembler;
+import org.eclipse.jdt.ls.core.internal.IDecompiler;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
+
+public class DecompilerManager {
+
+	public static final String DECOMPILED_HEADER = " // Failed to get sources. Instead, stub sources have been generated by the %s.";
+	public static final String DISASSEMBLER_HEADER_ADDITION = " // Implementation of methods is unavailable.\n";
+	private static final String LF = "\n";
+	private static final String EXTENSION_POINT_ID = "org.eclipse.jdt.ls.core.decompiler";
+	private static final String CLASS = "class";
+	private static final String ID = "id";
+	private static final String NAME = "name";
+
+	private final PreferenceManager preferenceManager;
+	private DecompilerDescriptor decompilerDescriptor;
+	private IDecompiler decompiler;
+
+	public DecompilerManager(PreferenceManager preferenceManager) {
+		this.preferenceManager = preferenceManager;
+	}
+
+	private static class DecompilerDescriptor {
+
+		private final IConfigurationElement configurationElement;
+		private IDecompiler decompiler;
+		public final String id;
+		public final String name;
+
+		public DecompilerDescriptor(IConfigurationElement element) {
+			configurationElement = element;
+			id = configurationElement.getAttribute(ID);
+			name = configurationElement.getAttribute(NAME);
+			decompiler = null;
+		}
+
+		public synchronized IDecompiler getDecompiler() {
+			if (decompiler == null) {

```

Comment:
I'm not sure the decompiler should be cached. That'd force extensions to make their decompilers as singletons. I don't want to make any assumption at that point. 
Unless there are performance concerns (which will need to be proven), there's no need to keep a decompiler instance in memory